### PR TITLE
boards: pinmux: stm32: configure serial pins from DT for STM32

### LINKED ISSUE #28934: pinmux: stm32: port remaining pinctrl DT serial definitions for STM32 based boards
**Is your enhancement proposal related to a problem? Please describe.**
The new pinmux configuration method introduced in #25996 means all of the STM32 based boards need to be updated to the new method.

For this we need to remove the USART/LPUART/UART definitions from **pinmux.c** file and add the corresponding pinctrl nodes to the boards' device tree.

**Describe the solution you'd like**
An ideal solution would be a script that executes the change automatically. Else we'll have to do it by and.
Difficult parts are to identify the correct pinctrl dtsi to use and then extracting the pin configuration from current pinmux.c.

**Describe alternatives you've considered**
For now, I have made a script that identifies every stm32 based boards and if it has defined pinmux for **lpuart/uart/usart** DT nodes. In which case, we'll need to update them manually.

Here is the script:
```bash
#!/bin/bash

pinmuxes=`find boards/arm/ -name "pinmux.c"`

#use Perl regex here to match with grep
#match for specific STM32 pinmux header, only used
#with ST based boards
match_stm32_pinmux="pinmux/stm32/pinmux_stm32.h"
match_peripheral_mux="(lp)*u(s)*art\d+"

for pinmux in $pinmuxes; do

	is_stm32_pinmux=`cat $pinmux | grep -cP $match_stm32_pinmux`
	if [ $is_stm32_pinmux -eq 0 ]
	then
		continue
	fi

	periph_mux_count=`cat $pinmux | grep -cP $match_peripheral_mux`
	periph_matches=`cat $pinmux | grep -oP $match_peripheral_mux`
	if [ $periph_mux_count -ne 0 ]
	then
		board=`dirname $pinmux`
		echo $board : $periph_matches
	fi
done
```

To this date, here are the boards that still need to upgrade to the new paradigm:
```
* boards/arm/nucleo_f302r8 : usart1 usart2 usart3
* boards/arm/nucleo_l552ze_q : lpuart1
* boards/arm/stm32f746g_disco : usart1 usart6
* boards/arm/stm32f723e_disco : usart2 usart6
* boards/arm/nucleo_f756zg : usart2 usart3 usart6
* boards/arm/nucleo_f334r8 : usart1 usart2 usart3
* boards/arm/black_f407zg_pro : usart1 usart2
* boards/arm/nucleo_f091rc : usart1 usart2
* boards/arm/dragino_lsn50 : usart1 usart2
* boards/arm/olimexino_stm32 : usart1 usart2 usart3
* boards/arm/stm32vl_disco : usart1 usart2 usart3
* boards/arm/96b_stm32_sensor_mez : usart1 usart2 usart3 uart4
* boards/arm/nucleo_f401re : usart1 usart2
* boards/arm/google_kukui : usart1
* boards/arm/nucleo_l4r5zi : usart1 lpuart1 usart2 usart3
* boards/arm/nucleo_f767zi : usart2 usart3 usart6
* boards/arm/nucleo_f413zh : usart3 usart6
* boards/arm/nucleo_g431rb : usart1 lpuart1
* boards/arm/nucleo_l073rz : usart1 usart2
* boards/arm/stm32f411e_disco : usart2
* boards/arm/stm32l476g_disco : usart2
* boards/arm/waveshare_open103z : usart1 usart2
* boards/arm/olimex_stm32_h407 : usart2 usart3 usart6
* boards/arm/nucleo_f207zg : usart3 usart6
* boards/arm/stm32f072_eval : usart2
* boards/arm/stm32l496g_disco : usart1 usart2 lpuart1
* boards/arm/black_f407ve : usart1 usart2
* boards/arm/96b_neonkey : usart1
* boards/arm/nucleo_l053r8 : usart1 usart2
* boards/arm/96b_aerocore2 : usart1 usart2 usart3 uart7 uart8
* boards/arm/nucleo_f070rb : usart1 usart2
* boards/arm/nucleo_h745zi_q : usart3 uart8
* boards/arm/stm32_min_dev : usart1 usart2 usart3
* boards/arm/stm32f030_demo : usart1
* boards/arm/stm32f429i_disc1 : usart1 usart2
* boards/arm/olimex_stm32_p405 : usart2
* boards/arm/stm32f412g_disco : usart2
* boards/arm/stm32f072b_disco : usart1
* boards/arm/olimex_stm32_h103 : usart1 usart2 usart3
* boards/arm/nucleo_g474re : usart1 lpuart1
* boards/arm/96b_wistrio : usart1 usart2
* boards/arm/stm3210c_eval : usart2
* boards/arm/stm32h747i_disco : usart1 uart8
* boards/arm/nucleo_l452re : usart1 usart2
* boards/arm/nucleo_l476rg : usart1 usart2 usart3
* boards/arm/blackpill_f411ce : usart1
* boards/arm/disco_l475_iot1 : uart4
* boards/arm/nucleo_f746zg : usart2 usart3 usart6
* boards/arm/nucleo_g071rb : usart2
* boards/arm/nucleo_f103rb : usart1 usart3
* boards/arm/stm32f4_disco : usart1 usart2
* boards/arm/96b_argonkey : usart1
* boards/arm/b_l4s5i_iot01a : usart1 uart4
* boards/arm/sensortile_box : usart1 usart2
* boards/arm/stm32f3_disco : usart1 usart2
* boards/arm/stm32f769i_disco : usart1 usart6
* boards/arm/stm32f469i_disco : usart3 usart6
* boards/arm/b_l072z_lrwan1 : usart1 usart2
* boards/arm/stm32f0_disco : usart1 usart2
* boards/arm/stm32l1_disco : usart1 usart2 usart3
* boards/arm/mikroe_mini_m4_for_stm32 : usart2
* boards/arm/stm32mp157c_dk2 : usart3 uart7
* boards/arm/steval_fcu001v1 : usart1
* boards/arm/nucleo_f030r8 : usart1 usart2
* boards/arm/stm32373c_eval : usart2
* boards/arm/adafruit_feather_stm32f405 : usart3
* boards/arm/nucleo_f429zi : usart3 usart6
* boards/arm/96b_avenger96 : uart4 uart7
* boards/arm/olimex_stm32_e407 : usart1 usart3 usart6
* boards/arm/stm32g0316_disco : usart1 usart1
* boards/arm/nucleo_f412zg : usart3 usart6
* boards/arm/nucleo_wb55rg : usart1 lpuart1
* boards/arm/nucleo_f446re : usart1 usart2
* boards/arm/nucleo_l496zg : usart2 lpuart1
* boards/arm/nucleo_f303re : usart2
* boards/arm/96b_carbon : usart1 usart2 usart6
* boards/arm/nucleo_l152re : usart2
* boards/arm/nucleo_l432kc : usart1 usart2
* boards/arm/nucleo_f411re : usart1 usart2
```
'*' PR providing new configuration is ready for these boards

#28851 should bring support for nucleo_l432kc ->closed moved to #28954 
#28954 ports all of the remaining boards
#28940 brings support for nucleo_h745zi_q

### PR DESCRIPTION
**Pull request by lochej**
Fixes #28934 by porting the remaining boards to the new pinctrl DT definitions.

### CODE REVIEW

**Review by gmarull — APPROVED** [review]
LGTM

**Comment by erwango** [comment]
> Fixes #28934 by porting the remaining boards to the new pinctrl DT definitions.

@lochej To be sure I understand the purpose of this PR, Do you plan to convert _all_ remaining boards with this PR ?

**Comment by lochej** [comment]
> 
> 
> > Fixes #28934 by porting the remaining boards to the new pinctrl DT definitions.
> 
> @lochej To be sure I understand the purpose of this PR, Do you plan to convert _all_ remaining boards with this PR ?

Absolutely 😄

File: boards/arm/stm32g0316_disco/pinmux.c
Comment (lochej) [inline boards/arm/stm32g0316_disco/pinmux.c:16]: @erwango what should I do for this one ? Looks like it is remapping PA11 to PA9 in **pinmux_stm32_init** when usart1 is used.

**Review by lochej — COMMENTED** [review]
I'm confused about G0 remapping function called for stm32g0316_disco board. Need some help for this one ! 🙃

File: boards/arm/stm32g0316_disco/pinmux.c
Comment (lochej) [reply boards/arm/stm32g0316_disco/pinmux.c:16]: image
@erwango @ABOSTM 
If I understand correctly, the G031 (8 pins) has AF **usart1_tx=pa9** and **usart1_rx=pb7**, but only **usart1_cts=pa11** is available on the 8 pins package, so you can remap pa11 to point on pa9 and get access to **usart1_tx=pa9** on pin 5 of the SO8 package. This indeed is necessary if we want the UART to work.

How are we going to handle this very specific case ?

**Comment by Nukersson** [comment]
b13fb76 looks good to me. I can provide approve only to some h7 SoC currently, sorry.

**Review by KozhinovAlexander — APPROVED** [review]

**Comment by lochej** [comment]
Fixed issues shown by buildkite in last rebase. Only 3 boards were to blame: 
* nucloe_f767zi : typo, missing a semicolon in the DT
* stm32f746g_disco : typo in usart6_rx_pc7 definition
* stm32l1_disco : bad dtsi  inclusion

File: .buildkite/pipeline.yml
Comment (lochej) [reply .buildkite/pipeline.yml:9]: This will be removed just before merging

File: boards/arm/stm32g0316_disco/pinmux.c
Comment (lochej) [reply boards/arm/stm32g0316_disco/pinmux.c:16]: @gmarull @erwango @ABOSTM I suggest we fix this one in another PR and just port to the new DT definitions as I did, but this specific remapping might need to be handled in the future.

File: boards/arm/nucleo_f756zg/nucleo_f756zg.dts
Comment (erwango) [inline boards/arm/nucleo_f756zg/nucleo_f756zg.dts:60]: Just a suggestion, but for readability (highly subjective) could have been:
```suggestion
	pinctrl-0 = <&usart2_tx_pd5 &usart2_rx_pd6
		     &usart2_rts_pd4 &usart2_cts_pd3>;
```

File: boards/arm/stm32f746g_disco/pinmux.c
Comment (erwango) [inline boards/arm/stm32f746g_disco/pinmux.c:17]: You're mixing 2 boards here

File: boards/arm/96b_wistrio/96b_wistrio.dts
Comment (erwango) [inline boards/arm/96b_wistrio/96b_wistrio.dts:49]: Based on pinmux.c file it would be:
```suggestion
	pinctrl-0 = <&usart2_tx_pa2 &usart2_rx_pa15>;
```
But maybe you fixed it on purpose ?

File: boards/arm/blackpill_f411ce/blackpill_f411ce.dts
Comment (erwango) [inline boards/arm/blackpill_f411ce/blackpill_f411ce.dts:9]:
```suggestion
#include <st/f4/stm32f411c(e-g)ux-pinctrl.dtsi>
``` ?

File: boards/arm/nucleo_f756zg/nucleo_f756zg.dts
Comment (lochej) [reply boards/arm/nucleo_f756zg/nucleo_f756zg.dts:60]: Agreed, this looks better, will apply this. If I remembre correctly, some other boards have same definitions, I will update them too.

File: boards/arm/stm32f746g_disco/pinmux.c
Comment (lochej) [reply boards/arm/stm32f746g_disco/pinmux.c:17]: I'm afraid you're going to have to elaborate a bit more, I don't understand where the mixing happens ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/96b_wistrio/96b_wistrio.dts
Comment (lochej) [reply boards/arm/96b_wistrio/96b_wistrio.dts:49]: Yes, **usart2_tx_pa15** is not a valid option for stm32l151cb-a.

File: boards/arm/nucleo_l053r8/nucleo_l053r8.dts
Comment (lochej) [reply boards/arm/nucleo_l053r8/nucleo_l053r8.dts:9]: Thanks !

File: boards/arm/blackpill_f411ce/blackpill_f411ce.dts
Comment (lochej) [reply boards/arm/blackpill_f411ce/blackpill_f411ce.dts:9]: Good catch thanks.

File: boards/arm/nucleo_f756zg/nucleo_f756zg.dts
Comment (lochej) [reply boards/arm/nucleo_f756zg/nucleo_f756zg.dts:60]: So, nucleo_f746zg is another board that will receive the same refactoring

Comment (lochej) [reply boards/arm/nucleo_f756zg/nucleo_f756zg.dts]: Done in latest push

File: boards/arm/stm32f746g_disco/pinmux.c
Comment (lochej) [reply boards/arm/stm32f746g_disco/pinmux.c:17]: Done in latest push

File: boards/arm/nucleo_l053r8/nucleo_l053r8.dts
Comment (lochej) [reply boards/arm/nucleo_l053r8/nucleo_l053r8.dts]: Done in latest push

File: boards/arm/blackpill_f411ce/blackpill_f411ce.dts
Comment (lochej) [reply boards/arm/blackpill_f411ce/blackpill_f411ce.dts]: done in latest push

**Comment by erwango** [comment]
@galak, should we add `pcintrl-name` prop ?
It's not used for now, but I think this is going in the direction we discussed yesterday.
This being said, it would be easy to add in batch once it is read to be used.

**Review by sidcha — APPROVED** [review]

**Comment by galak** [comment]
Looks like nucleo_l4r5zi fails.

I recommend to do the following test locally:

`./scripts/sanitycheck -T samples/hello_world/ --cmake-only -l -a arm`

**Comment by galak** [comment]
Now that 28858 will need to rebase this PR as well.

**Comment by lochej** [comment]
> Now that #28858 will need to rebase this PR as well.

Thanks ! I get to work 😄

**Comment by lochej** [comment]
> Looks like nucleo_l4r5zi fails.
> 
> I recommend to do the following test locally:
> 
> `./scripts/sanitycheck -T samples/hello_world/ --cmake-only -l -a arm`

Here you go, build passed on my computer:
```
dev@ubuntu:~/zephyrproject/zephyr$ ./scripts/sanitycheck -T samples/hello_world/ --cmake-only -l -a arm
INFO    - JOBS: 8
INFO    - Selecting all possible platforms per test case
INFO    - Building initial testcase list...
INFO    - 292 test configurations selected, 52 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:  240/ 240  100%  skipped:    0, failed:    0
INFO    - 0 of 240 tests passed (0.00%), 0 failed, 52 skipped with 0 warnings in 137.26 seconds
INFO    - In total 240 test cases were executed on 292 out of total 293 platforms (99.66%)
INFO    - 0 tests executed on platforms, 240 tests were only built.
```

**Comment by galak** [comment]
@lochej what about nucleo_h745zi_q?

**Comment by lochej** [comment]
> @lochej what about nucleo_h745zi_q?

PR #28940 It is actually provided and tested by @Nukersson 
@Nukersson would you like your PR  #28940 commits to be added to this list instead or would you like it to be merged seperately ?

**Comment by galak** [comment]
> PR #28940 It is actually provided and tested by @Nukersson
> @Nukersson would you like your PR #28940 commits to be added to this list instead or would you like it to be merged seperately ?
> 
> Doesn't matter for me. Please take changes, which are faster to integrate.

Ah, thanks forgot about that PR.  Lets leave both PRs as is.  Was waiting for @erwango to approve #28940 before merging.

**Review by erwango — APPROVED** [review]

**Review by Navin-Sankar — APPROVED** [review]

**Comment by lochej** [comment]
@galak We should be ready for merging 😄 You'll just have to drop the buildkite commit when merging and we'll be done with UARTs.

**Comment by galak** [comment]
passed CI, so removing bk commit and merging.

**Comment by galak** [comment]
> @galak We should be ready for merging 😄 You'll just have to drop the buildkite commit when merging and we'll be done with UARTs.

Yep, just did it.

**Pull request merged**
